### PR TITLE
feat: add DNS and TCP reports to test-connectivity

### PR DESCRIPTION
**Pull request by fortuna**
@amircybersec Check out this implementation.

I am able to get both TCP and DNS reports, while keeping the main test agnostic to the transport. (Thanks to your httptrace trick!)

image

Thoughts?

### CODE REVIEW

**Comment by fortuna** [comment]
I added timing for DNS, but not TCP yet. It requires a map of net-address to start time.

**Comment by amircybersec** [comment]
@fortuna I like the proposed approach. It basically makes the context specific to a given dialer which can solve the concurrency issue and provide access to hostname. We can technically refactor this into a `reporter dialer` that can be used as the baseDialer. We can use the existing approach in complement to this and add `do53-udp, do53-tcp, doh, dot, https, quic, etc` test reports on the top of this. I am going to take this code for a spin 🚘

**Comment by amircybersec** [comment]
@fortuna I made a light refactor. I also added start time and duration to TCP connection. 

Also, it seems like your branch is on the top of an older version that does not have the recent udp support for socks. I can do a rebase. 
```
2024/09/05 22:03:29 Failed to create PacketDialer: config scheme 'socks5' is not supported for Packet Dialers
````

**Comment by fortuna** [comment]
I've moved your changes to https://github.com/Jigsaw-Code/outline-sdk/pull/275.

I'm not touching any dependencies in my PR, so if it doesn't have UDP SOCKS5, it's because main doesn't have it.

**Comment by fortuna** [comment]
Oh, nevermind, I started in an old main

**Comment by fortuna** [comment]
Ok, my branch is not on top of main @ HEAD

**Review by fortuna — COMMENTED** [review]
Let me know if you have any comments on this

**Comment by fortuna** [comment]
Now with colored slog!
image

File: x/examples/test-connectivity/main.go [hunk 1]
```diff
@@ -161,56 +211,141 @@ func main() {
...
+							Time:       dnsStart.UTC().Truncate(time.Second),
+							DurationMs: time.Since(dnsStart).Milliseconds(),
+						}
+						if di.Err != nil {
+							report.Error = di.Err.Error()
+						}
+						for _, ip := range di.Addrs {
+							report.AnswerIPs = append(report.AnswerIPs, ip.IP.String())
+						}
+						// TODO(fortuna): Use a Mutex.
```
Comment (amircybersec) [inline x/examples/test-connectivity/main.go]:
```suggestion
mu.Lock()
```

File: x/examples/test-connectivity/main.go [hunk 2]
```diff
@@ -161,56 +211,141 @@ func main() {
...
+						dnsReports = append(dnsReports, report)
+					},
```
Comment (amircybersec) [inline x/examples/test-connectivity/main.go]:
```suggestion
mu.Unlock()
					},
```

**Review by amircybersec — COMMENTED** [review]
please add mutex lock/unlock for `dnsReport` in UDP too. Other than that, it looks good. I will do a separate PR to add `connectStart` and `Duration` to `tcpReport` after we merge this.

File: x/examples/test-connectivity/main.go (hunk 1 above)
Comment (fortuna) [reply x/examples/test-connectivity/main.go]: Done in https://github.com/Jigsaw-Code/outline-sdk/pull/272/commits/acdc62201c16952ebbf4176759a2d8c9f33fa8a8

File: x/examples/test-connectivity/main.go (hunk 2 above)
Comment (fortuna) [reply x/examples/test-connectivity/main.go]: Done in https://github.com/Jigsaw-Code/outline-sdk/pull/272/commits/acdc62201c16952ebbf4176759a2d8c9f33fa8a8

File: x/examples/test-connectivity/main.go [hunk 3]
```diff
@@ -161,21 +207,94 @@ func main() {
...
 		resolverAddress := net.JoinHostPort(resolverHost, "53")
 		for _, proto := range strings.Split(*protoFlag, ",") {
 			proto = strings.TrimSpace(proto)
 			var resolver dns.Resolver
+			dnsReports := make([]dnsReport, 0)
+			tcpReports := make([]tcpReport, 0)
 			switch proto {
 			case "tcp":
+				configToDialer := config.NewDefaultConfigToDialer()
+				configToDialer.BaseStreamDialer = transport.FuncStreamDialer(func(ctx context.Context, addr string) (transport.StreamConn, error) {
```
[the remark this answers is not in the document]
Comment (fortuna) [reply x/examples/test-connectivity/main.go]: Done in https://github.com/Jigsaw-Code/outline-sdk/pull/272/commits/acdc62201c16952ebbf4176759a2d8c9f33fa8a8

**Review by amircybersec — APPROVED** [review]

**Pull request merged**